pilot/pkg/serviceregistry/external/servicediscovery.go

```diff
@@ -93,51 +93,67 @@ func NewServiceDiscovery(configController model.ConfigStoreCache, store model.Is
 		configController.RegisterEventHandler(serviceEntryKind,
 			func(old, curr model.Config, event model.Event) {
 				cs := convertServices(curr)
+				names := map[string]struct{}{}
 
 				// If it is add/delete event we should always do a full push. If it is update event, we should do full push,
 				// only when services have changed - otherwise, just push endpoint updates.
-				fp := true
-				if event == model.EventUpdate {
+				var addedSvcs, deletedSvcs, updatedSvcs, unchangedSvcs []*model.Service
+
+				switch event {
+				case model.EventUpdate:
 					// This is not needed, update should always have old populated, but just in case.
 					if old.Spec != nil {
 						os := convertServices(old)
-						fp = servicesChanged(os, cs)
+						addedSvcs, deletedSvcs, updatedSvcs, unchangedSvcs = servicesDiff(cs, os)
 					} else {
+						unchangedSvcs = cs
 						log.Warnf("Spec is not available in the old service entry during update, proceeding with full push %v", old)
 					}
-				}
-
-				// If service entry is deleted, cleanup endpoint shards for services.
-				if event == model.EventDelete {
+				case model.EventDelete:
+					deletedSvcs = cs
+					// If service entry is deleted, cleanup endpoint shards for services.
 					for _, svc := range cs {
 						c.XdsUpdater.SvcUpdate(c.Cluster(), string(svc.Hostname), svc.Attributes.Namespace, event)
 					}
+				case model.EventAdd:
+					addedSvcs = cs
+				default:
+					unchangedSvcs = cs
+				}
+
+				for _, svc := range addedSvcs {
+					names[string(svc.Hostname)] = struct{}{}
 				}
+				for _, svc := range deletedSvcs {
+					names[string(svc.Hostname)] = struct{}{}
+				}
+				for _, svc := range updatedSvcs {
+					names[string(svc.Hostname)] = struct{}{}
+				}
+
+				willFullPush := len(names) > 0
+				refreshIndexes := willFullPush // If will trigger full-push, lazy-update instance index.
```

Comment:
Does the same logic (add/delete/updated) exist on k8s side already?

Reply:
I do not see that. Despite on that I think from the view of istio we should know the services changes from ServiceEntry change other than the serviceEntry changes in order to archieve the full-push scoping.